File: tunnel-common/src/main/java/com/tuuzed/tunnel/common/proto/ProtoMessageDecoder.java
package com.tuuzed.tunnel.common.proto;

import com.tuuzed.tunnel.common.proto.internal.ProtoMessageCodec;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.LengthFieldBasedFrameDecoder;

public class ProtoMessageDecoder extends LengthFieldBasedFrameDecoder {

    public ProtoMessageDecoder() {
        super(2 * 1024 * 1024, 0, ProtoMessageCodec.MESSAGE_FRAME_FIELD_SIZE, 0, 0);
    }

    @Override
    protected Object decode(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
        final ByteBuf in2 = (ByteBuf) super.decode(ctx, in);
        final ProtoMessage message = ProtoMessageCodec.decode(in2);
        if (in2 != null) {
            in2.release();
        }
        return message;
    }
}

File: tunnel2-common/src/main/kotlin/tunnel2/common/proto/ProtoMessageDecoder.kt
package tunnel2.common.proto

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.LengthFieldBasedFrameDecoder

class ProtoMessageDecoder : LengthFieldBasedFrameDecoder(
    4 * 1024 * 1024,
    0,
    /* lengthFieldLength */  ProtoMessage.MESSAGE_FRAME_FIELD_SIZE,
    0,
    0,
    true
) {
    override fun decode(ctx: ChannelHandlerContext?, inbuf: ByteBuf?): Any? {
        return ProtoMessage.decode(inbuf)
    }
}